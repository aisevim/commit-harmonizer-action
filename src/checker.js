import { rulesConfig } from './rules.js';

function createMatrix(rows, cols) {
  return Array.from({ length: rows }, () => Array.from({ length: cols }, () => " "));
}

function applyRuleToMatrix(_matrix, position, errorMessage) {
	const matrix = structuredClone(_matrix);

  for (let i = 0; i < matrix.length; i++) {
    const line = matrix[i];
    const middlePosition = position[0] + Math.floor((position[1] - position[0]) / 2);

    if (i === 0) {
			line[middlePosition] = "↑";
		} else {
			line[middlePosition] = i + 1 === matrix.length ? "╵" : '┆';
		}

    if (i === matrix.length - 1) {
      line[middlePosition + 1] = '-';
      line[middlePosition + 2] = '-';
      line[middlePosition + 3] = '-';
      line[middlePosition + 5] = errorMessage;
    }
  }

  return matrix;
}

function formatLog(matrix) {
  return matrix.map(row => row.join('').replace(/\s+$/g, '')).join('\n');
}

function hasError(matrix) {
	return Boolean(matrix.flat(2).join('').replace(/\s+$/g, ''))
}

export function checkCommitMessages(message) {
  const maxLength = message.length + 10;
  let matrix = createMatrix(3, maxLength);

  for (const rule of rulesConfig) {
    const match = message.match(rule.regex);
    const position = match?.indices?.groups?.position;

    if (!position) {
			continue
		}

		matrix = applyRuleToMatrix(matrix, position, rule.errorMessage);
		matrix.push(...createMatrix(2, maxLength));
  }

  return [hasError(matrix), `\n${message}\n${formatLog(matrix)}`];
}
