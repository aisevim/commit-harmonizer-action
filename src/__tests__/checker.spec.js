import { expect, test, describe } from 'vitest'
import { checkCommitMessages } from '../checker'

describe('Generated Logs', () => {
  test("1 match on no description", () => {
		const [, log] = checkCommitMessages(`docs : Update documentation`);
		expect(log).toMatchSnapshot()
  })

	test("2 match on no description", () => {
		const [, log] = checkCommitMessages(`docs(): `);
		expect(log).toMatchSnapshot()
  })

	test("3 match on no description", () => {
		const [, log] = checkCommitMessages(`docs() : `);
		expect(log).toMatchSnapshot()
  })

	test("4 match on no description", () => {
		const [, log] = checkCommitMessages(`docs () : `);
		expect(log).toMatchSnapshot()
  })

	test("5 match on no description", () => {
		const [, log] = checkCommitMessages(`docs () : `);
		expect(log).toMatchSnapshot()
  })

	test("colon missing", () => {
		const [, log] = checkCommitMessages(`Update documentation`);
		expect(log).toMatchSnapshot()
  })

	test("space before", () => {
		const [, log] = checkCommitMessages(` docs: Update documentation`);
		expect(log).toMatchSnapshot()

		const [, log2] = checkCommitMessages(` docs:`);
		expect(log2).toMatchSnapshot()
  })
})

describe('Has Errors', () => {
  test("return true on error", () => {
		const [hasError] = checkCommitMessages(`docs: `);
    expect(hasError).toBeTruthy()

		const [hasError2] = checkCommitMessages(``);
    expect(hasError2).toBeTruthy()
  })

	test("return false on valid", () => {
		const [hasError] = checkCommitMessages(`docs: fooo`);
    expect(hasError).toBeFalsy()
  })
})
