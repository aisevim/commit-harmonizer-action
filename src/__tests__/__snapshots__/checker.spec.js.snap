// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generated Logs > 1 match on no description 1`] = `
"
docs : Update documentation
    ↑
    ┆
    ╵--- Unexpected character between commit type and colon.

"
`;

exports[`Generated Logs > 2 match on no description 1`] = `
"
docs(): 
     ↑ ↑
     ┆ ┆
     ┆ ╵--- Missing or empty commit description.
     ┆
     ╵--- Empty commit scope provided.

"
`;

exports[`Generated Logs > 3 match on no description 1`] = `
"
docs() : 
     ↑↑ ↑
     ┆┆ ┆
     ┆┆ ╵--- Missing or empty commit description.
     ┆┆
     ┆╵--- Unexpected character before the colon in the commit message.
     ┆
     ╵--- Empty commit scope provided.

"
`;

exports[`Generated Logs > 4 match on no description 1`] = `
"
docs () : 
    ↑ ↑↑ ↑
    ┆ ┆┆ ┆
    ┆ ┆┆ ╵--- Missing or empty commit description.
    ┆ ┆┆
    ┆ ┆╵--- Unexpected character before the colon in the commit message.
    ┆ ┆
    ┆ ╵--- Empty commit scope provided.
    ┆
    ╵--- Unexpected character between commit type and scope.

"
`;

exports[`Generated Logs > 5 match on no description 1`] = `
"
docs () :
    ↑ ↑↑ ↑
    ┆ ┆┆ ┆
    ┆ ┆┆ ┆--- Missing or empty commit description.
    ┆ ┆┆ ┆
    ┆ ┆┆ ╵--- Space after the colon is required.
    ┆ ┆┆
    ┆ ┆╵--- Unexpected character before the colon in the commit message.
    ┆ ┆
    ┆ ╵--- Empty commit scope provided.
    ┆
    ╵--- Unexpected character between commit type and scope.

"
`;

exports[`Generated Logs > colon missing 1`] = `
"
Update documentation
                    ↑
                    ┆
                    ╵--- Colon is missing in the commit message structure.

"
`;

exports[`Generated Logs > space before 1`] = `
"
 docs: Update documentation
↑
┆
╵--- Avoid leading or trailing spaces in the commit message.

"
`;

exports[`Generated Logs > space before 2`] = `
"
 docs:
↑
┆
╵--- Avoid leading or trailing spaces in the commit message.

"
`;
